File: ORMConvertor/Tests/Differential/ResultRow.cs
using System.Globalization;

namespace Tests.Differential;

/// <summary>
/// The canonical text form of one row of a query result (decision 089). Both ecosystems
/// render into it and the comparison is of the rendered text, because a pair whose halves
/// run in different runtimes cannot meet anywhere else - the decision rejected an endpoint
/// that would run foreign code for exactly that purpose.
///
/// Every rule below has a counterpart in <c>ResultRow.java</c>, and that the two agree is
/// not assumed: <c>renderer-conformance.txt</c> is rendered by both suites and compared
/// against the same checked-in text. It is the first thing to run, because everything else
/// here rests on it.
///
/// A type the rules do not name throws rather than guessing. Guessing is what the whole
/// decision is against: a value rendered by a rule nobody wrote would make two suites agree
/// or disagree for a reason no one could read out of the file.
/// </summary>
internal static class ResultRow
{
    /// <summary>What a null field renders as. No string can collide with it: strings are quoted.</summary>
    public const string Null = "NULL";

    /// <summary>Separator between fields. A tab inside a string value is escaped, so it cannot occur unescaped.</summary>
    public const char Separator = '\t';

    /// <summary>
    /// How much of a value survives rendering. It is a property of what the query computes -
    /// an average over money bears a different scale than a plain column - so it travels with
    /// the query in the matrix, not with the host the suite runs on (decision 089).
    /// </summary>
    /// <param name="DecimalScale">Decimal places an exact decimal is rounded to, half to even.</param>
    /// <param name="FloatSignificantDigits">
    /// Significant digits an approximate number is rounded to. Binary floating point differs
    /// between the ecosystems in the last bits, and insisting on those would measure the
    /// driver rather than the translation.
    /// </param>
    /// <param name="FractionalSecondDigits">Fractional digits of a timestamp, as the mapping states them.</param>
    /// <remarks>
    /// A record class and not a record struct, and that is not a matter of taste: on a
    /// struct <c>new()</c> is the zero-initializing constructor, so the defaults written
    /// here would never run and every caller who did not spell all three out would quietly
    /// render at scale zero.
    /// </remarks>
    public sealed record RenderSettings(
        int DecimalScale = 6,
        int FloatSignificantDigits = 12,
        int FractionalSecondDigits = 3);

    /// <summary>Renders one row: its fields in the order the matrix states, joined by the separator.</summary>
    public static string Render(IReadOnlyList<object?> fields, RenderSettings settings)
        => string.Join(Separator, fields.Select(field => RenderField(field, settings)));

    /// <summary>
    /// Orders rendered rows the way a query without an ordering instruction has to be
    /// compared: as a set, by the rendered line itself. Ordinal, because the Java side
    /// compares UTF-16 code units and nothing else would agree with it.
    /// </summary>
    public static List<string> Sorted(IEnumerable<string> rows)
        => [.. rows.OrderBy(row => row, StringComparer.Ordinal)];

    public static string RenderField(object? value, RenderSettings settings) => value switch
    {
        null or DBNull => Null,

        bool flag => flag ? "true" : "false",

        // Every integral width renders the same way: an integer has no facets to lose.
        sbyte or byte or short or ushort or int or uint or long or ulong
            => Convert.ToInt64(value, CultureInfo.InvariantCulture).ToString(CultureInfo.InvariantCulture),

        decimal exact => RenderDecimal(exact, settings.DecimalScale),

        float approximate => RenderApproximate(approximate, settings.FloatSignificantDigits),
        double approximate => RenderApproximate(approximate, settings.FloatSignificantDigits),

        DateTime timestamp => RenderTimestamp(timestamp, settings.FractionalSecondDigits),
        DateOnly date => date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture),

        string text => RenderString(text),

        _ => throw new NotSupportedException(
            $"A value of type {value.GetType().FullName} has no rendering rule in decision 089, "
            + "so the canonical form cannot state it. Add the rule to both suites and to the "
            + "conformance file, or keep the type out of the matrix - do not let this guess."),
    };

    /// <summary>
    /// An exact decimal at the stated scale, rounded half to even. The scale is stated rather
    /// than taken from the value because the same column arrives with different scales from
    /// different drivers, and the comparison must not depend on that.
    /// </summary>
    private static string RenderDecimal(decimal value, int scale)
        => Math.Round(value, scale, MidpointRounding.ToEven)
            .ToString("F" + scale.ToString(CultureInfo.InvariantCulture), CultureInfo.InvariantCulture);

    /// <summary>
    /// An approximate number rounded to the stated significant digits and then written
    /// without an exponent and without trailing zeros. The last step is what makes the two
    /// ecosystems agree: one of them would otherwise write 3 and the other 3.0.
    /// </summary>
    private static string RenderApproximate(double value, int significantDigits)
    {
        if (double.IsNaN(value) || double.IsInfinity(value))
        {
            throw new NotSupportedException(
                $"The value {value} is not a number the canonical form can state (decision 089).");
        }

        decimal exact;
        try
        {
            exact = (decimal)value;
        }
        catch (OverflowException)
        {
            throw new NotSupportedException(
                $"The value {value} is outside the range the canonical form renders (decision 089).");
        }

        if (exact != 0m)
        {
            // Significant digits, not decimal places: the exponent of the value decides how
            // many places that is. Clamped to what decimal rounding accepts at either end.
            var exponent = (int)Math.Floor(Math.Log10((double)Math.Abs(exact)));
            var places = Math.Clamp(significantDigits - 1 - exponent, 0, 28);
            exact = Math.Round(exact, places, MidpointRounding.ToEven);
        }

        return exact.ToString("0.############################", CultureInfo.InvariantCulture);
    }

    private static string RenderTimestamp(DateTime value, int fractionalDigits)
    {
        var fraction = fractionalDigits > 0
            ? "." + new string('f', fractionalDigits)
            : string.Empty;

        return value.ToString("yyyy-MM-dd'T'HH:mm:ss" + fraction, CultureInfo.InvariantCulture);
    }

    /// <summary>
    /// A string in quotes, with the four characters escaped that would otherwise make one
    /// field look like two, or one row like two.
    /// </summary>
    private static string RenderString(string value)
    {
        var text = new System.Text.StringBuilder(value.Length + 2);
        text.Append('"');

        foreach (var character in value)
        {
            switch (character)
            {
                case '"': text.Append("\\\""); break;
                case '\\': text.Append("\\\\"); break;
                case '\t': text.Append("\\t"); break;
                case '\n': text.Append("\\n"); break;
                case '\r': text.Append("\\r"); break;
                default: text.Append(character); break;
            }
        }

        text.Append('"');
        return text.ToString();
    }
}

File: ORMConvertor/JavaTests/src/test/java/cz/stochel/ormconvertor/javatests/differential/ResultRow.java
package cz.stochel.ormconvertor.javatests.differential;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.sql.Timestamp;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

/**
 * The canonical text form of one row of a query result (decision 089). The counterpart of
 * {@code ResultRow.cs}, rule for rule: both ecosystems render into this form and the
 * comparison is of the rendered text, because the two halves of a pair run in different
 * runtimes and cannot meet anywhere else.
 *
 * <p>That the two agree is not assumed. {@code renderer-conformance.txt} is rendered by
 * both suites against the same checked-in text, and it is the first thing to run: BigDecimal
 * and decimal print trailing zeros differently, LocalDateTime and DateTime format midnight
 * differently, and a tab inside a string breaks the separator if one side forgets to escape
 * it.
 *
 * <p>A type no rule names throws rather than guessing. A value rendered by a rule nobody
 * wrote would make the two suites agree or disagree for a reason no one could read out of
 * the file.
 */
public final class ResultRow {

    /** What a null field renders as. No string can collide with it: strings are quoted. */
    public static final String NULL = "NULL";

    /** Separator between fields. A tab inside a string is escaped, so it never occurs bare. */
    public static final char SEPARATOR = '\t';

    private ResultRow() {
    }

    /**
     * How much of a value survives rendering. It travels with the query in the matrix and
     * not with the host the suite runs on: an average over money bears a different scale
     * than a plain column (decision 089).
     */
    public record Settings(int decimalScale, int floatSignificantDigits, int fractionalSecondDigits) {

        public static Settings defaults() {
            return new Settings(6, 12, 3);
        }
    }

    /** One row: its fields in the order the matrix states, joined by the separator. */
    public static String render(List<Object> fields, Settings settings) {
        StringBuilder text = new StringBuilder();

        for (int index = 0; index < fields.size(); index++) {
            if (index > 0) {
                text.append(SEPARATOR);
            }
            text.append(renderField(fields.get(index), settings));
        }

        return text.toString();
    }

    /**
     * Orders rendered rows the way a query without an ordering instruction has to be
     * compared: as a set, by the rendered line itself. The .NET side sorts ordinally, which
     * is the same UTF-16 code unit order {@link String#compareTo} gives.
     */
    public static List<String> sorted(List<String> rows) {
        List<String> ordered = new ArrayList<>(rows);
        ordered.sort(String::compareTo);
        return ordered;
    }

    public static String renderField(Object value, Settings settings) {
        if (value == null) {
            return NULL;
        }

        if (value instanceof Boolean flag) {
            return flag ? "true" : "false";
        }

        // Every integral width renders the same way: an integer has no facets to lose.
        if (value instanceof Byte || value instanceof Short || value instanceof Integer || value instanceof Long) {
            return Long.toString(((Number) value).longValue());
        }

        if (value instanceof BigDecimal exact) {
            return exact.setScale(settings.decimalScale(), RoundingMode.HALF_EVEN).toPlainString();
        }

        if (value instanceof Float || value instanceof Double) {
            return renderApproximate(((Number) value).doubleValue(), settings.floatSignificantDigits());
        }

        if (value instanceof Timestamp timestamp) {
            return renderTimestamp(timestamp.toLocalDateTime(), settings.fractionalSecondDigits());
        }

        if (value instanceof LocalDateTime timestamp) {
            return renderTimestamp(timestamp, settings.fractionalSecondDigits());
        }

        if (value instanceof java.sql.Date date) {
            return date.toLocalDate().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
        }

        if (value instanceof LocalDate date) {
            return date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
        }

        if (value instanceof String text) {
            return renderString(text);
        }

        throw new UnsupportedOperationException(
                "A value of type " + value.getClass().getName() + " has no rendering rule in decision 089, "
                        + "so the canonical form cannot state it. Add the rule to both suites and to the "
                        + "conformance file, or keep the type out of the matrix - do not let this guess.");
    }

    /**
     * An approximate number rounded to the stated significant digits and then written
     * without an exponent and without trailing zeros. The last step is what makes the two
     * ecosystems agree: one of them would otherwise write 3 and the other 3.0.
     */
    private static String renderApproximate(double value, int significantDigits) {
        if (Double.isNaN(value) || Double.isInfinite(value)) {
            throw new UnsupportedOperationException(
                    "The value " + value + " is not a number the canonical form can state (decision 089).");
        }

        return BigDecimal.valueOf(value)
                .round(new MathContext(significantDigits, RoundingMode.HALF_EVEN))
                .stripTrailingZeros()
                .toPlainString();
    }

    private static String renderTimestamp(LocalDateTime value, int fractionalDigits) {
        String fraction = fractionalDigits > 0 ? "." + "S".repeat(fractionalDigits) : "";
        return value.format(DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss" + fraction));
    }

    /**
     * A string in quotes, with the four characters escaped that would otherwise make one
     * field look like two, or one row like two.
     */
    private static String renderString(String value) {
        StringBuilder text = new StringBuilder(value.length() + 2);
        text.append('"');

        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            switch (character) {
                case '"' -> text.append("\\\"");
                case '\\' -> text.append("\\\\");
                case '\t' -> text.append("\\t");
                case '\n' -> text.append("\\n");
                case '\r' -> text.append("\\r");
                default -> text.append(character);
            }
        }

        text.append('"');
        return text.toString();
    }
}
